### PR TITLE
Document APIs for the System.Composition.Convention Namespace

### LINKED ISSUE #2261: Add Undocumented APIs: System.Composition.Convention Namespace
Add missing documentation for the System.Composition.Convention Namespace

This namespace contains the following classes:

ConventionBuilder 
ExportConventionBuilder 
ImportConventionBuilder 
ParameterImportConventionBuilder 
PartConventionBuilder 
PartConventionBuilder<T> 
AttributedModelProvider  (This is a base class of ConventionBuilder and will be documented.)

The namespace topic docs/xml/ns-System.Composition.Convention.xml will also be documented.

Information will be pulled from the source code at:
https://github.com/dotnet/corefx/tree/master/src/System.Composition.Convention/src/System/Composition/Convention

Note:
Missing information is marked by the string: "To be added."

This issue is part of a larger project to fill in information for undocumented APIs.

@mairaw

### PR DESCRIPTION
**Pull request by tompratt-AQ**
# Title
Document APIs for the System.Composition.Convention Namespace

## Summary
Added documentation for "To be added." fields in all classes of the System.Composition.Convention Namespace:

- ConventionBuilder 
- ExportConventionBuilder 
- ImportConventionBuilder 
- ParameterImportConventionBuilder 
- PartConventionBuilder 
- PartConventionBuilderOfT 

**Note:** I will be adding two more files for this namespace in another commit:
The AttributedModelProvider base class, and the namespace file ns-System.Composition.Convention.xml

Fixes #2261 

## Suggested Reviewers
@rpetrusha

### CODE REVIEW

File: xml/System.Composition.Convention/ImportConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/ImportConventionBuilder.xml]: no comma after contract

File: xml/System.Composition.Convention/ParameterImportConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/ParameterImportConventionBuilder.xml]: Move "This class is used..." to the Remarks section. Also, a semicolon, not a comma, should be used.

File: xml/System.Composition.Convention/PartConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder.xml]: calculates --> determines

File: xml/System.Composition.Convention/PartConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder.xml]: that will be exported --> to export

File: xml/System.Composition.Convention/PartConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder.xml]: that will be exported --> to export

File: xml/System.Composition.Convention/PartConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder.xml]: Filter --> A filter

File: xml/System.Composition.Convention/PartConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder.xml]: which of the available constructors will be used --> which constructor is used

File: xml/System.Composition.Convention/PartConventionBuilder.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder.xml]: which of the available constructors will be used --> which constructor is used

**Review by rpetrusha — CHANGES_REQUESTED** [review]
The namespace also needs to be documented. The namespace also contains a type named AttributedModelProvider that needs to be documented.

**Comment by tompratt-AQ** [comment]
@rpetrusha Thanks for the detailed review. I'm working on the namespace topic and the base class topic right now.

**Comment by tompratt-AQ** [comment]
@rpetrusha I added AttributedModelProvider.xml and ns-System.Composition.Convention.xml for you to review, and made the suggested changes from your last review. Thanks!

File: xml/System.Composition.Convention/AttributedModelProvider.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/AttributedModelProvider.xml]: This can be a bit clearer: Serves as the base class that defines the core behavior of methods used to retrieve metadata attributes for a type.

File: xml/System.Composition.Convention/AttributedModelProvider.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/AttributedModelProvider.xml]: Generally, we just want to document protected constructors as, "Creates a new instance of the <whatever> class."
In the Remarks, you can add a comment that this is a protected constructor that can only be called from derived classes.

**Comment by tompratt-AQ** [comment]
@rpetrusha Could you please review the Exceptions that I added in this last commit to docs/xml/System.Composition.Convention/PartConventionBuilder`1.xml.  
They are thrown by internal methods that are called by the public methods to which I added them. They are custom Argument exceptions for this class, the code for them is in:  https://github.com/dotnet/corefx/blob/master/src/System.Composition.Convention/src/System/Composition/ExceptionBuilder.cs
They are the only exceptions in the namespace.  Thanks!

**Comment by tompratt-AQ** [comment]
@mairaw Does this PR need to be restarted? It passed all checks last Friday and is waiting for a second review of a couple items (see my previous comment), but as of yesterday it has been stuck on these new additional checks.

File: xml/System.Composition.Convention/PartConventionBuilder`1.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder`1.xml]: `<paramref>` is a general formatting tag for use in text. So the exception description should read, "The `<paramref name="propertySelector" />` expression must be a `<see cref="T:System.Linq.Expressions.MemberExpression"/>` for accessing a property." That's true of each of the uses of the paramref element here.

Also, in each case, "a" should precede the type name ("a `<see cref="T:System.Linq.Expressions.MemberExpression"/>`"). There are a few cases where you've done that, but it should be done in all cases.

File: xml/System.Composition.Convention/PartConventionBuilder`1.xml
Comment (rpetrusha) [inline xml/System.Composition.Convention/PartConventionBuilder`1.xml]: Instead of "expression must be new", "expression must use the `new` operator"

**Review by rpetrusha — COMMENTED** [review]
@tompratt-AQ, two comments, but they apply to all of the exception items added.

**Comment by tompratt-AQ** [comment]
@rpetrusha Thanks for reviewing the exceptions!

**Comment by rpetrusha** [comment]
Thanks for making the changes, @tompratt-AQ. This looks good. We should have someone from the PU review this before we merge, though. I'll try to find out who owns it.

**Comment by rpetrusha** [comment]
@tompratt-AQ, it turns out that this and the other 2 undocumented MEF namespaces (System.Composition.Hosting and System.Composition.Hosting.Core) were documented in CAPS as OOB releases and, because they weren't part of MREF, weren't migrated to GitHub. Please take a look at them to make sure that we've provided as much detail as the old documentation. The documentation is accessible on MSDN from the [System.Composition namespaces](https://msdn.microsoft.com/en-us/library/windows/apps/jj126278.aspx) topic.

**Comment by tompratt-AQ** [comment]
@rpetrusha There is a 4th namespace involved as well, that I just started working on: System.Composition
I will check all of them against MSDN. 
**Question:**  Should I change all documentation for these to the MSDN wording? Or just evaluate the two, modify when they don't match, and add any missing information?

**Comment by rpetrusha** [comment]
@tompratt-AQ, evaluate when they don't match and add missing information. They don't have to be identical; select the version that makes the most sense.

**Comment by tompratt-AQ** [comment]
@rpetrusha I checked the documentation against the [MSDN topics](https://msdn.microsoft.com/en-us/library/windows/apps/jj126278.aspx) and made adjustments accordingly.

**Comment by mairaw** [comment]
@safern @AlexGhiondea do you own this API area? Can you guys help review it?
This content is available to review on staging at https://review.docs.microsoft.com/en-us/dotnet/api/system.composition.convention?branch=pr-en-us-2309&view=netcore-2.0

/cc @karelz 

Thanks!

**Comment by mairaw** [comment]
Sorry about that @safern. I read the wrong line at https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#areas 😄

**Comment by rpetrusha** [comment]
@tompratt-AQ, since this largely duplicates content from MSDN with some improvements, I think it's safe to merge.

**Review by rpetrusha — APPROVED** [review]

**Review by AlexGhiondea — COMMENTED** [review]
Thanks for doing this @tompratt-AQ !

**Pull request merged**
